catalog/migrate/iterator.go

```diff
@@ -0,0 +1,152 @@
+package migrate
+
+import (
+	"context"
+	"errors"
+
+	"github.com/treeverse/lakefs/catalog"
+	"github.com/treeverse/lakefs/catalog/mvcc"
+	"github.com/treeverse/lakefs/catalog/rocks"
+	"github.com/treeverse/lakefs/db"
+)
+
+type iteratorState int
+
+const (
+	iteratorStateInit iteratorState = iota
+	iteratorStateQuerying
+	iteratorStateDone
+)
+
+type Iterator struct {
+	state     iteratorState
+	ctx       context.Context
+	db        db.Database
+	branchID  int64
+	commitID  int64
+	buf       []*rocks.EntryRecord
+	err       error
+	offset    string
+	fetchSize int
+	value     *rocks.EntryRecord
+}
+
+var ErrIteratorClosed = errors.New("iterator closed")
+
+// NewIterator returns an iterator over an mvcc branch/commit, giving entries as EntryCatalog.
+func NewIterator(ctx context.Context, db db.Database, branchID int64, commitID int64, fetchSize int) *Iterator {
+	return &Iterator{
+		ctx:       ctx,
+		db:        db,
+		branchID:  branchID,
+		commitID:  commitID,
+		buf:       make([]*rocks.EntryRecord, 0, fetchSize),
+		fetchSize: fetchSize,
+	}
+}
+
+func (it *Iterator) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	it.maybeFetch()
+
+	// stage a value and increment offset
+	if len(it.buf) == 0 {
+		return false
+	}
+	it.value = it.buf[0]
+	it.offset = string(it.value.Path)
+	if len(it.buf) > 1 {
+		it.buf = it.buf[1:]
+	} else {
+		it.buf = it.buf[:0]
+	}
```

Comment:
We know `len(it.buf) >= 1`, so isn't this just
```suggestion
	it.buf = it.buf[1:]
```
:question: 

I can see no performance advantage to `it.buf = it.buf[:0]`, given that by the time `len(it.buf)==1` that slice is anyway a _subslice_ of the allocated slice.

Reply:
totally right, I'll align all the rest of the code that uses this way.